katharsis-client/src/main/java/io/katharsis/client/internal/BaseResponseDeserializer.java

```diff
@@ -0,0 +1,198 @@
+
+import io.katharsis.dispatcher.controller.resource.ResourceUpsert;
+import io.katharsis.queryParams.QueryParams;
+import io.katharsis.repository.RepositoryMethodParameterProvider;
+import io.katharsis.request.dto.DataBody;
+import io.katharsis.request.dto.RequestBody;
+import io.katharsis.request.path.JsonPath;
+import io.katharsis.resource.information.ResourceInformation;
+import io.katharsis.resource.registry.RegistryEntry;
+import io.katharsis.resource.registry.ResourceRegistry;
+import io.katharsis.response.BaseResponseContext;
+import io.katharsis.response.CollectionResponseContext;
+import io.katharsis.response.JsonApiResponse;
+import io.katharsis.response.ResourceResponseContext;
+import io.katharsis.utils.parser.TypeParser;
+
+/**
+ * Deerializes top-level JSON object and provides ability to include compound documents
+ * 
+ * TODO could this be merged with RequestBodyDeserializer?
+ */
+public class BaseResponseDeserializer extends JsonDeserializer<BaseResponseContext> {
+
+	private static final String INCLUDED_FIELD_NAME = "included";
+	private static final String DATA_FIELD_NAME = "data";
+	private static final String META_FIELD_NAME = "meta";
+	private static final String LINKS_FIELD_NAME = "links";
+
+	private ResourceRegistry resourceRegistry;
+	private ObjectMapper objectMapper;
+	
+	private TypeParser typeParser = new TypeParser();
+
+	public BaseResponseDeserializer(ResourceRegistry resourceRegistry, ObjectMapper objectMapper) {
+		this.resourceRegistry = resourceRegistry;
+		this.objectMapper = objectMapper;
+	}
+
+	@Override
+	public BaseResponseContext deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException, JsonProcessingException {
+
+		JsonNode node = jp.readValueAsTree();
+		if (node == null) {
+			return null;
+		}
+
+		JsonNode data = node.get(DATA_FIELD_NAME);
+		JsonNode included = node.get(INCLUDED_FIELD_NAME);
+		if(data == null){
+			throw new IllegalStateException("no data received");
+		}
+		
+		ClientResourceUpsert upsert = new ClientResourceUpsert(resourceRegistry, typeParser, objectMapper);
+		
+		ResourceBodies dataBodies = upsert.parse(data, jp);
+		ResourceBodies includedBodies = upsert.parse(included, jp);
+		
+		upsert.allocateResources(dataBodies);
+		upsert.allocateResources(includedBodies);
+		
+		upsert.setRelations(dataBodies);
+		upsert.setRelations(includedBodies);
+
+		// FIXME get linkage
+		if (dataBodies.isCollection) {
+			JsonApiResponse response = new JsonApiResponse();
+			response.setEntity(dataBodies.resources);
+			CollectionResponseContext collectionResponse = new CollectionResponseContext(response, null, null);
+
+			return collectionResponse;
+		} else {
+			JsonApiResponse apiResponse = new JsonApiResponse();
+
+			if (dataBodies.resources.size() == 1) {
+				apiResponse.setEntity(dataBodies.resources.get(0));
+			} else {
+				apiResponse.setEntity(null);
+			}
+
+			ResourceResponseContext resourceResponse = new ResourceResponseContext(apiResponse, -1);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Immediately return this expression instead of assigning it to the temporary variable "resourceResponse". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1488)